<a href="https://colab.research.google.com/github/kshamie3505-a11y/Codveda-Internship-Tasks/blob/main/Level3Task2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Natural Language Processing (NLP) - Text Classification**
**Description:** Classify text data into categories (e.g., spam vs. non-spam, sentiment analysis). **Objectives** Preprocess text data (tokenization, removing stopwords, stemming/lemmatization). Convert text into numerical representation using TF-IDF or Word2Vec. Train a classification model (e.g., Naive Bayes, Logistic Regression) on the processed text. Evaluate the model using precision, recall, and F1-score. Tools: Python, nltk, scikit-learn, pandas.

In [3]:
import pandas as pd
df = pd.read_csv('/content/emails.csv')
df

,text,spam
0,Subject: naturally irresistible your corporate...,1
1,Subject: the stock trading gunslinger fanny i...,1
2,Subject: unbelievable new homes made easy im ...,1
3,Subject: 4 color printing special request add...,1
4,"Subject: do not have money , get software cds ...",1
...,...,...
5723,Subject: re : research and development charges...,0
5724,"Subject: re : receipts from visit jim , than...",0
5725,Subject: re : enron case study update wow ! a...,0
5726,"Subject: re : interest david , please , call...",0


In [4]:
df.info

<bound method DataFrame.info of                                                    text  spam
0     Subject: naturally irresistible your corporate...     1
1     Subject: the stock trading gunslinger  fanny i...     1
2     Subject: unbelievable new homes made easy  im ...     1
3     Subject: 4 color printing special  request add...     1
4     Subject: do not have money , get software cds ...     1
...                                                 ...   ...
5723  Subject: re : research and development charges...     0
5724  Subject: re : receipts from visit  jim ,  than...     0
5725  Subject: re : enron case study update  wow ! a...     0
5726  Subject: re : interest  david ,  please , call...     0
5727  Subject: news : aurora 5 . 2 update  aurora ve...     0

[5728 rows x 2 columns]>

In [5]:
df.columns


Index(['text', 'spam'], dtype='object')

In [6]:
df.shape

(5728, 2)

In [7]:
df.isnull().sum()

,0
text,0
spam,0


**Preprocess Text Data**

Performing essential text preprocessing steps using nltk. This includes tokenizing the text, removing common stopwords, and applying stemming or lemmatization to normalize words. This step will prepare the text for feature extraction, focusing on the 'text' column and using the 'spam' column as labels.


In [8]:
# Importing the necessary libraries like NLTK
import nltk
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

print("NLTK modules and resources loaded successfully.")

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


NLTK modules and resources loaded successfully.


**Instantiating the WordNetLemmatizer:** Defining a text preprocessing function that includes tokenization, stopword removal, and lemmatization, and then apply this function to the 'text' column of the DataFrame to create a new 'preprocessed_text' column.



In [9]:
nltk.download('punkt_tab', quiet=True)

lemmatizer = WordNetLemmatizer()
stop_words = set(stopwords.words('english'))

def preprocess_text(text):
    text = text.lower()
    tokens = word_tokenize(text)
    filtered_tokens = [word for word in tokens if word.isalpha() and word not in stop_words]
    lemmatized_tokens = [lemmatizer.lemmatize(word) for word in filtered_tokens]
    return " ".join(lemmatized_tokens)

df['preprocessed_text'] = df['text'].apply(preprocess_text)
print("Text preprocessing complete. A new column 'preprocessed_text' has been added to the DataFrame.")

Text preprocessing complete. A new column 'preprocessed_text' has been added to the DataFrame.


**Convert Text to Numerical Features**
Transforming the preprocessed text into numerical features suitable for machine learning using scikit-learn's TF-IDF Vectorizer. This will convert text documents into a matrix of TF-IDF features.


In [10]:
# Importing TfidfVectorizer from sklearn
from sklearn.feature_extraction.text import TfidfVectorizer
print("TfidfVectorizer imported successfully.")

TfidfVectorizer imported successfully.


**Fitting and transforming the preprocessed_text column of the DataFrame into a TF-IDF matrix, storing it in a variable named X.**



In [11]:
vectorizer = TfidfVectorizer()
X = vectorizer.fit_transform(df['preprocessed_text'])
print("TF-IDF vectorization complete. The TF-IDF matrix is stored in 'X'.")
print(f"Shape of TF-IDF matrix (X): {X.shape}")

TF-IDF vectorization complete. The TF-IDF matrix is stored in 'X'.
Shape of TF-IDF matrix (X): (5728, 30960)


**Training the Classification Model**

split the numerical data into training and testing sets. Then, train a classification model, such as Logistic Regression from scikit-learn, on the training data to learn the patterns for text classification.


In [12]:
# Importing the necessary modules train_test_split and LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

print("Modules 'train_test_split' and 'LogisticRegression' imported successfully.")

Modules 'train_test_split' and 'LogisticRegression' imported successfully.


In [13]:
y = df['spam']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LogisticRegression(max_iter=1000) # Increased max_iter for convergence
model.fit(X_train, y_train)

print("Data split into training and testing sets successfully.")
print(f"Shape of X_train: {X_train.shape}")
print(f"Shape of X_test: {X_test.shape}")
print("Logistic Regression model trained successfully.")

Data split into training and testing sets successfully.
Shape of X_train: (4582, 30960)
Shape of X_test: (1146, 30960)
Logistic Regression model trained successfully.


**Evaluating the Model Performance**: using key metrics like precision,recall and F1-score to understand how well the model classifies text




In [14]:
# Importing the necessary metrics from sklearn.metrics
from sklearn.metrics import classification_report, accuracy_score
print("Modules 'classification_report' and 'accuracy_score' imported successfully.")

Modules 'classification_report' and 'accuracy_score' imported successfully.


**Using the trained model to make predictions on the test set, and then print the classification report and accuracy score to evaluate the model's performance.**



In [15]:
y_pred = model.predict(X_test)

print("Classification Report:")
print(classification_report(y_test, y_pred))

print("\nAccuracy Score:")
print(accuracy_score(y_test, y_pred))


Classification Report:
              precision    recall  f1-score   support

           0       0.97      1.00      0.98       856
           1       0.99      0.90      0.94       290

    accuracy                           0.97      1146
   macro avg       0.98      0.95      0.96      1146
weighted avg       0.97      0.97      0.97      1146


Accuracy Score:
0.9729493891797557


**Summary:**

The NLP text classification process involved several key stages:

1.  **Data Preprocessing**: The 'text' column of the email dataset was cleaned and prepared using nltk. This included converting text to lowercase, tokenizing words, removing non-alphabetic characters and common English stopwords, and lemmatizing words to their base form. This resulted in a new preprocessed_text column.
2.  **Feature Extraction**: The preprocessed text was then converted into numerical features using scikit-learn's TfidfVectorizer. This transformed the text data into a TF-IDF matrix, making it suitable for machine learning algorithms.
3.  **Model Training**: The numerical features (X) and 'spam' labels (y) were split into training and testing sets (80% for training, 20% for testing). A Logistic Regression model was then trained on the training data.
4.  **Model Evaluation**: The trained Logistic Regression model's performance was evaluated on the test set using standard classification metrics: precision, recall, F1-score, and accuracy.

The model demonstrated strong performance in classifying emails as spam or not spam. It achieved an overall accuracy of approximately 0.973 on the test set. For non-spam (class 0), the model showed excellent performance with a precision of 0.97, recall of 1.00, and F1-score of 0.98. For spam (class 1), the model achieved a precision of 0.99, recall of 0.90, and an F1-score of 0.94. The high precision for spam indicates that when the model predicts an email is spam, it is almost always correct, while the slightly lower recall suggests it might miss a small percentage of actual spam emails.

**Data Analysis Key Findings**

*   Text preprocessing successfully cleaned the 'text' data by lowercasing, tokenizing, removing stopwords, and lemmatizing, creating a preprocessed_text column.
*   TF-IDF vectorization transformed the preprocessed text into a numerical matrix X with a shape of (5728, 30960), representing 5728 documents and 30960 unique features.
*   The dataset was split into training and testing sets, with X_train having a shape of (4582, 30960) and X_test having a shape of (1146, 30960).
*   The Logistic Regression model achieved an overall accuracy of approximately 0.973 on the test set.
*   For non-spam (class 0), the model had a precision of 0.97, a recall of 1.00, and an F1-score of 0.98.
*   For spam (class 1), the model had a high precision of 0.99, a recall of 0.90, and an F1-score of 0.94.

